### PR TITLE
IC-1866 add more fine grained control over ignored and unsuccessful requests in app insights

### PR DESCRIPTION
**Pull request by tomsmyers**
## What does this pull request do?

add more fine grained control over ignored and unsuccessful requests in app insights
 - allows requests matching patterns to be ignored
 - allows customization of which status codes are treated as errors in the UI

## What is the intent behind these changes?

make app insights more useable

### CODE REVIEW

File: server/azureAppInsights.ts [hunk 1]
```diff
@@ -4,14 +4,20 @@ import applicationVersion from './applicationVersion'
...
   envelope: Contracts.EnvelopeTelemetry,
   _contextObjects: { [name: string]: unknown } | undefined
 ): boolean {
   if (envelope.data.baseType === Contracts.TelemetryTypeString.Request) {
     const requestData = envelope.data.baseData
-    if (requestData instanceof Contracts.RequestData && requestData.name.startsWith('GET /assets/')) {
-      return false
+    if (requestData instanceof Contracts.RequestData) {
+      const { excludedRequests } = config.applicationInsights
+      for (let i = 0; i < excludedRequests.length; i += 1) {
```
Comment (Gweaton) [inline server/azureAppInsights.ts:15]: This doesn't really matter, and I might be missing something or could this be an ES6 `for...of`? Is the index important?

File: server/azureAppInsights.ts [hunk 2]
```diff
@@ -31,6 +37,21 @@ function addUsernameProcessor(
   return true
 }
 
+function errorStatusCodeProcessor(
+  envelope: Contracts.EnvelopeTelemetry,
+  _contextObjects: { [name: string]: unknown } | undefined
+): boolean {
+  if (envelope.data.baseType === Contracts.TelemetryTypeString.Request) {
+    if (envelope.data.baseData !== undefined) {
```
Comment (Gweaton) [inline server/azureAppInsights.ts:45]: can this be part of the above `if` as an `&&` condition so we don't have the nested `if`?

**Review by Gweaton — APPROVED** [review]
I'm not sure I understand all the config changes here, but have left a couple of pedantic JS comments - enjoy!

File: server/azureAppInsights.ts (hunk 1 above)
Comment (tomsmyers) [reply server/azureAppInsights.ts:15]: intellij really doesn't like you using that thing... the reasons are compelling but boring

File: server/azureAppInsights.ts (hunk 2 above)
Comment (tomsmyers) [reply server/azureAppInsights.ts:45]: yes 👍

File: server/azureAppInsights.ts [hunk 3]
```diff
@@ -25,12 +31,25 @@ function addUsernameProcessor(
...
     if (userId) {
       // eslint-disable-next-line no-param-reassign
-      envelope.tags['ai.user.authUserId'] = userId
+      envelope.tags[defaultClient.context.keys.userAuthUserId] = userId
     }
   }
   return true
 }
 
+// function errorStatusCodeProcessor(
```
Comment (sldblog) [inline server/azureAppInsights.ts:40]: Can these commented blocks be removed until they're ready to use?

**Review by sldblog — COMMENTED** [review]
So if I understand correctly, the core aim of this is to ignore requests that `startswith "GET /health"`?

👍 Thanks

**Pull request merged**